File: Limbus/limbus_core/frontend/Source.fs
module Source

open System.IO
open MessageIF
type  TCHAR =  C of char | EOL | EOF 

[<Class>]
type Source(reader : StreamReader )  = 
    let mutable rearch_eof : bool = false
    let mutable line = ""
    
    interface  MessageProducer with 
        member this.add_message_listener(ml: MessageListener) : unit = 1 |> ignore
        member this.remove_message_listener(ml : MessageListener) : unit = 1 |> ignore
        member this.send_message(m : Message) : unit = 1 |> ignore
    
    member this.reader = reader
    member val line_num = 0 with get, set
    member val current_pos = -2 with get, set
    
    member this.send_message(m : Message) : unit = 1 |> ignore
    
    member this.current_char() : TCHAR =
        let line_len = String.length line
        
        if rearch_eof 
        then EOF
        else 
            match this.line_num with 
                | -2 ->
                    // first time
                    this.read_line() 
                    this.next_char()
                | -1 -> EOL
                | _ when this.current_pos = line_len -> EOL
                | _ when this.current_pos > line_len ->
                    this.read_line() 
                    this.next_char()
                | _ -> C line.[this.current_pos]
        
    member this.read_line() : unit =
        line <- this.reader.ReadLine()
        this.current_pos <- 0

        if line = null then 
            rearch_eof <- true
            let m = {msg_type = PARSER_SUMMARY ; body = MSGS [string(this.line_num) ; line]}
            this.send_message(m)
            
        this.line_num <- this.line_num + 1
        
    
    member this.next_char() : TCHAR = 
        this.current_pos <- this.current_pos + 1
        this.current_char()
    
    member this.peek_char() : TCHAR = 
        this.current_char() |> ignore

        let next_pos = this.current_pos + 1        
        match rearch_eof with 
            | true -> EOF
            | false when next_pos >= String.length line -> EOL
            | false -> C line.[next_pos]
                
    member this.close() : unit =
        this.reader.Close()


        

File: Limbus/limbus_core/frontend/source.py
# -*- coding: utf-8 -*-
from .. message import MessageProducer, MessageHandler, Message, MessageType


class Source(MessageProducer):
    EOL = '\n'
    EOF = None

    def __init__(self, reader):
        self.line_num = 0
        self.current_pos = -2  # -2が最初の値
        self.reader = reader
        self.line = ""
        self.message_handler = MessageHandler()

    def current_char(self):
        # first line
        if self.current_pos == -2:
            self.read_line()
            return self.next_char()

        # eof?
        if self.line == self.EOF:
            return self.EOF

        if self.current_pos == len(self.line):
            return self.EOL

        if self.current_pos > len(self.line):
            self.read_line()
            return self.next_char()

        return self.line[self.current_pos]

    def read_line(self):
        self.line = self.reader.readline()

        if self.line == "":
            self.line = self.EOF
            return self.EOF

        self.line = self.line.rstrip('\n')

        self.current_pos = -1
        self.line_num = self.line_num + 1
        ln = self.line_num
        line = self.line
        msg = Message(MessageType.SOURCE_LINE, (ln, line))
        self.send_message(msg)

    def next_char(self):
        self.current_pos = self.current_pos + 1
        return self.current_char()

    def peek_char(self):
        self.current_char()
        if self.line == self.EOF:
            return self.EOF

        next_pos = self.current_pos + 1
        if next_pos > len(self.line):
            return self.line[next_pos]
        else:
            return self.EOL

    def close(self):
        self.reader.close()

    def get_line_num(self):
        return self.line_num

    def get_position(self):
        return self.current_pos

    # delegate
    def add_message_listener(self, listener):
        self.message_handler.add_message_listener(listener)

    def remove_message_listener(self, listener):
        self.message_handler.remove_message_listener(listener)

    def send_message(self, message):
        self.message_handler.send_message(message)
